pkg/gate/gate.go

```diff
@@ -9,56 +9,70 @@ import (
 
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promauto"
-	"github.com/prometheus/prometheus/pkg/gate"
+	promgate "github.com/prometheus/prometheus/pkg/gate"
 )
 
-type Gater interface {
-	IsMyTurn(ctx context.Context) error
+// Gate is an interface that mimics prometheus/pkg/gate behaviour.
+type Gate interface {
+	Start(ctx context.Context) error
 	Done()
 }
 
 // Gate wraps the Prometheus gate with extra metrics.
-type Gate struct {
-	g               *gate.Gate
+type gate struct {
+	g *promgate.Gate
+	m *metrics
+}
+
+type metrics struct {
 	inflightQueries prometheus.Gauge
 	gateTiming      prometheus.Histogram
 }
 
-// NewGate returns a new query gate.
-func NewGate(maxConcurrent int, reg prometheus.Registerer) *Gate {
-	g := &Gate{
-		g: gate.New(maxConcurrent),
-		inflightQueries: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
-			Name: "gate_queries_in_flight",
-			Help: "Number of queries that are currently in flight.",
-		}),
-		gateTiming: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
-			Name:    "gate_duration_seconds",
-			Help:    "How many seconds it took for queries to wait at the gate.",
-			Buckets: []float64{0.01, 0.1, 0.3, 0.6, 1, 3, 6, 9, 20, 30, 60, 90, 120, 240, 360, 720},
-		}),
+// Keeper keeps gate safe and sound while holding some metrics.
```

Comment:
Not sure I understand this comment. Maybe we could say `Keeper is used to create multiple gates sharing the same metrics.`?